// testo-example/kotlin-js.js
(function (factory) {
  if (typeof define === 'function' && define.amd)
    define(['exports', './kotlin-kotlin-stdlib.js'], factory);
  else if (typeof exports === 'object')
    factory(module.exports, require('./kotlin-kotlin-stdlib.js'));
  else {
    if (typeof globalThis['kotlin-kotlin-stdlib'] === 'undefined') {
      throw new Error("Error loading module 'kotlin-js'. Its dependency 'kotlin-kotlin-stdlib' was not found. Please, check whether 'kotlin-kotlin-stdlib' is loaded prior to 'kotlin-js'.");
    }
    globalThis['kotlin-js'] = factory(typeof globalThis['kotlin-js'] === 'undefined' ? {} : globalThis['kotlin-js'], globalThis['kotlin-kotlin-stdlib']);
  }
}(function (_, kotlin_kotlin) {
  'use strict';
  //region block: imports
  var constructor = (class IteratorYieldResult {
    static [Symbol.hasInstance](instance) {
        return instance && (typeof instance === 'object') && (instance.done === false)
    }
    }.prototype).constructor;
  var iterator = kotlin_kotlin.$_$.u4;
  var CoroutineImpl = kotlin_kotlin.$_$.h3;
  var Unit_instance = kotlin_kotlin.$_$.o1;
  var protoOf = kotlin_kotlin.$_$.i4;
  var THROW_CCE = kotlin_kotlin.$_$.y5;
  var SequenceScope = kotlin_kotlin.$_$.t4;
  var get_COROUTINE_SUSPENDED = kotlin_kotlin.$_$.g3;
  var initMetadataForLambda = kotlin_kotlin.$_$.z3;
  var VOID = kotlin_kotlin.$_$.c;
  //endregion
  //region block: pre-declaration
  initMetadataForLambda(iteratorFor$slambda, CoroutineImpl, VOID, [1]);
  //endregion
  function get_POW() {
    _init_properties_BigInt_kt__agomkx();
    return POW;
  }
  var POW;
  function get_n(_this__u8e3s4) {
    _init_properties_BigInt_kt__agomkx();
    return BigInt(_this__u8e3s4);
  }
  var properties_initialized_BigInt_kt_nhzooh;
  function _init_properties_BigInt_kt__agomkx() {
    if (!properties_initialized_BigInt_kt_nhzooh) {
      properties_initialized_BigInt_kt_nhzooh = true;
      POW = Function('base', 'exponent', 'return base ** exponent');
    }
  }
  function invoke(_this__u8e3s4, p1, p2) {
    // Inline function 'kotlin.js.asDynamic' call
    return _this__u8e3s4(p1, p2);
  }
  function iteratorFor(source) {
    // Inline function 'js.iterable.JsIteratorLike.iterator' call
    var this_0 = source[Symbol.iterator]();
    return iteratorFor_0(this_0);
  }
  function iteratorFor_0(source) {
    return iterator(iteratorFor$slambda_0(source, null));
  }
  function iteratorFor$slambda($source, resultContinuation) {
    this.cf_1 = $source;
    CoroutineImpl.call(this, resultContinuation);
  }
  protoOf(iteratorFor$slambda).hf = function ($this$iterator, $completion) {
    var tmp = this.if($this$iterator, $completion);
    tmp.g8_1 = Unit_instance;
    tmp.h8_1 = null;
    return tmp.m8();
  };
  protoOf(iteratorFor$slambda).d9 = function (p1, $completion) {
    return this.hf(p1 instanceof SequenceScope ? p1 : THROW_CCE(), $completion);
  };
  protoOf(iteratorFor$slambda).m8 = function () {
    var suspendResult = this.g8_1;
    $sm: do
      try {
        var tmp = this.e8_1;
        switch (tmp) {
          case 0:
            this.f8_1 = 5;
            this.e8_1 = 1;
            continue $sm;
          case 1:
            var tmp_0 = this;
            var tmp_1 = this.cf_1.next();
            tmp_0.ef_1 = tmp_1 instanceof constructor ? tmp_1 : null;
            if (this.ef_1 == null) {
              this.e8_1 = 4;
              var tmp_2 = this;
              continue $sm;
            } else {
              this.ff_1 = this.ef_1;
              this.e8_1 = 2;
              continue $sm;
            }

          case 2:
            this.gf_1 = this.ff_1;
            this.e8_1 = 3;
            suspendResult = this.df_1.bc(this.gf_1.value, this);
            if (suspendResult === get_COROUTINE_SUSPENDED()) {
              return suspendResult;
            }

            continue $sm;
          case 3:
            this.e8_1 = 1;
            continue $sm;
          case 4:
            return Unit_instance;
          case 5:
            throw this.h8_1;
        }
      } catch ($p) {
        var e = $p;
        if (this.f8_1 === 5) {
          throw e;
        } else {
          this.e8_1 = this.f8_1;
          this.h8_1 = e;
        }
      }
     while (true);
  };
  protoOf(iteratorFor$slambda).if = function ($this$iterator, completion) {
    var i = new iteratorFor$slambda(this.cf_1, completion);
    i.df_1 = $this$iterator;
    return i;
  };
  function iteratorFor$slambda_0($source, resultContinuation) {
    var i = new iteratorFor$slambda($source, resultContinuation);
    var l = function ($this$iterator, $completion) {
      return i.hf($this$iterator, $completion);
    };
    l.$arity = 1;
    return l;
  }
  //region block: exports
  _.$_$ = _.$_$ || {};
  _.$_$.a = iteratorFor_0;
  _.$_$.b = iteratorFor;
  //endregion
  return _;
}));
